File: backend/src/routes/exitPrices.ts
import { Router } from 'express';
import type { ExitPrices } from '../types';

export const exitPricesRouter = Router();

// GET /api/exit-prices — returns { coinId: exitPrice } for the user
exitPricesRouter.get('/', async (req, res) => {
  const { data, error } = await req.supabase!.from('exit_prices').select('coin_id, exit_price');

  if (error) {
    res.status(500).json({ error: error.message });
    return;
  }

  const result: ExitPrices = {};
  for (const row of data as { coin_id: string; exit_price: number }[]) {
    result[row.coin_id] = Number(row.exit_price);
  }
  res.json(result);
});

// PUT /api/exit-prices — body: { coinId, exitPrice }
exitPricesRouter.put('/', async (req, res) => {
  const { coinId, exitPrice } = req.body as { coinId?: string; exitPrice?: number };
  if (!coinId || typeof exitPrice !== 'number') {
    res.status(400).json({ error: 'coinId and exitPrice are required.' });
    return;
  }

  if (exitPrice <= 0) {
    const { error } = await req
      .supabase!.from('exit_prices')
      .delete()
      .eq('coin_id', coinId)
      .eq('user_id', req.userId!);
    if (error) {
      res.status(500).json({ error: error.message });
      return;
    }
    res.status(204).send();
    return;
  }

  const { error } = await req.supabase!.from('exit_prices').upsert({
    user_id: req.userId!,
    coin_id: coinId,
    exit_price: exitPrice,
  });

  if (error) {
    res.status(500).json({ error: error.message });
    return;
  }
  res.status(204).send();
});

File: backend/app/routes/exit_prices.py
from fastapi import APIRouter, Depends, HTTPException, Response, status
from app.dependencies import AuthContext, require_auth
from app.db.postgres_client import get_conn
from app.models import ExitPriceUpdate

router = APIRouter()


@router.get("/", response_model=dict[str, float])
def get_exit_prices(auth: AuthContext = Depends(require_auth)):
    conn = get_conn()
    try:
        with conn.cursor() as cur:
            cur.execute(
                "SELECT coin_id, exit_price FROM exit_prices WHERE user_id = %s",
                (auth.user_id,),
            )
            return {row["coin_id"]: float(row["exit_price"]) for row in cur.fetchall()}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.put("/", status_code=status.HTTP_204_NO_CONTENT)
def set_exit_price(body: ExitPriceUpdate, auth: AuthContext = Depends(require_auth)):
    conn = get_conn()
    try:
        with conn.cursor() as cur:
            if body.exitPrice <= 0:
                cur.execute(
                    "DELETE FROM exit_prices WHERE user_id = %s AND coin_id = %s",
                    (auth.user_id, body.coinId),
                )
            else:
                cur.execute(
                    "INSERT INTO exit_prices (user_id, coin_id, exit_price)"
                    " VALUES (%s, %s, %s)"
                    " ON CONFLICT (user_id, coin_id) DO UPDATE SET exit_price = EXCLUDED.exit_price,"
                    " updated_at = now()",
                    (auth.user_id, body.coinId, body.exitPrice),
                )
        conn.commit()
    except Exception as e:
        conn.rollback()
        raise HTTPException(status_code=500, detail=str(e))
    return Response(status_code=status.HTTP_204_NO_CONTENT)
